src/source-af-packet.c

```diff
@@ -2118,6 +2266,193 @@ TmEcode AFPSetBPFFilter(AFPThreadVars *ptv)
     return TM_ECODE_OK;
 }
 
+#ifdef HAVE_PACKET_EBPF
+static int AFPInsertHalfFlow(int mapd, void *key, uint64_t inittime)
+{
+    unsigned int nr_cpus = UtilCpuGetNumProcessorsConfigured();
+    struct pair value[nr_cpus];
+    unsigned int i;
+
+    if (mapd == -1) {
+        return 0;
+    }
+
+    for (i = 0; i < nr_cpus; i++) {
+        value[i].time = inittime;
+        value[i].packets = 0;
+        value[i].bytes = 0;
+    }
+    SCLogDebug("Inserting element in eBPF mapping: %lu", inittime);
+    if (bpf_map_update_elem(mapd, key, value, BPF_NOEXIST) != 0) {
+        switch (errno) {
+            case E2BIG:
+            case EEXIST:
+                return 0;
+            default:
+                SCLogError(SC_ERR_BPF, "Can't update eBPF map: %s (%d)",
+                        strerror(errno),
+                        errno);
+                return 0;
+        }
+    }
+    return 1;
+}
+#endif
+
+static int AFPBypassCallback(Packet *p)
+{
+#ifdef HAVE_PACKET_EBPF
+    SCLogDebug("Calling af_packet callback function");
+    /* Only bypass TCP and UDP */
+    if (!(PKT_IS_TCP(p) || PKT_IS_UDP(p))) {
+        return 0;
+    }
+
+    /* Bypassing tunneled packets is currently not supported
+     * because we can't discard the inner packet only due to
+     * primitive parsing in eBPF */
+    if (IS_TUNNEL_PKT(p)) {
+        return 0;
+    }
+    struct timespec curtime;
+    uint64_t inittime = 0;
+    if (clock_gettime(CLOCK_MONOTONIC, &curtime) == 0) {
```

Comment:
why is this getting the time while the packet has a timestamp as well? comment

Reply:
This is because the time is the monotonic clock (the time since start of the computer). In the timestamp of the packet we have a usual one.